Run as service

leonowski opened the pull request:
So, I've been trying to work out a way to make this run similar to the Windows version of Blue Iris.  I'm not sure if this is even a good approach, but a side effect of this effort is that Blue Iris now uses the correct time.  The "fix" was to import a windows registry reg file that I used to try and set up the service automagically (along with specifying the correct TZ env var for the container).  It does seem to work as there are 2 separate BlueIris.exe processes running.  Unfortunately, when I close the GUI process, it seems to also close the service.  I'm still working on that.

Because of the way services work in Windows, I thought that it may be possible for the container to run continuously even if it is in a bad state.  To try and get around that, I created a small "check_process.sh" script that runs every 30 seconds.  If it doesn't find any BlueIris.exe processes, the script will send a kill to PID 1.  I disabled automatic restart for the blueiris.sh script in supervisord so that this check_process.sh script can handle this.  Since PID 1 is being killed in the container, the container now needs the `--init` option turned on.  For docker-compose, the option is documented [here ](https://docs.docker.com/compose/compose-file/#init).

I haven't explored the possibility of the service requiring a Blue Iris license to run.  I have also observed, in my testing, the service running without the UI - I just haven't figure it all out yet.  I wanted to get this PR in to discuss anyway.

There are also some startup improvements in this:

- Visual C++ 2019 runtime is installed quietly before the blueiris.exe installer.  There still is a Wine Mono and Wine Gecko installer dialog that requires manual interaction.  I'm working to find the best way to get rid of that.

- After install, PID 1 will be killed so a fresh restart brings up BlueIris.exe as a service.

Lastly, I'm still having trouble with recordings on the 64-bit version.  Can anyone else confirm that it is a problem for them?  For me, my Blue Iris will crash when I attempt to record anything.  This doesn't seem to happen on the 32-bit version.

leonowski committed:
try to run blueiris as windows service

leonowski committed:
remove my test URL

TonyBrobston commented:
@leonowski I'll try running `leonowski/docker-blueiris:run_as_service` for a bit. I do have a license. Do I have to go into Blue Iris settings and check the box to run as a service or does it do it automatically? 

Also, if you haven't purchased a license, I did find a discount code on the interwebz; I think it was 15% off. Totaled to like $66. I figured even if I don't run Blue Iris long term; $66 is worth the peace of mind of the same question you're asking, "Does the free version have less functionality?"

As for the 32-bit vs. 64-bit. I did have the Blue Iris crash on 64-bit, but it happened randomly; when I click record it does fine. I have had less issues on 32-bit, but I still have had a few issues. Like right now, `8080` is up and running fine, but `81` randomly stopped serving. I'm also seeing a lot of `004e:fixme:msvcrt:MSVCRT__stdio_common_vswprintf options 26 not handled` and `004e:fixme:msvcrt:MSVCRT__stdio_common_vsnwprintf_s options 24 not handled` in the logs. I'm not sure where I should put the details since I'm running on `leonowski/docker-blueiris` currently.

TonyBrobston commented:
@leonowski When I check the box to run Blue Iris as a service, it asked for a username/password. I assume I don't have to give it one? After that it told me to restart Blue Iris. I was thinking I could just hit the power button in the top right and it would automatically restart with the script that you said runs everything 30 seconds? It didn't seem to ever restart, so I did a `docker restart` on the container. When it comes back up I get this error:
![image](https://user-images.githubusercontent.com/4724577/92135557-87782580-edd0-11ea-8a78-7a389ee6dae4.png)

This puts me in a pretty sore state as I can't open Blue Iris to turn the "run as service" off. I'm not sure how to recover other than to blow my setup away and start from scratch. In the future I'll be sure to swap my volume mount to point to like `/tmp` and test before modifying my real setup.

TonyBrobston commented:
When I got into the container and ran the `blueiris.sh`, I got some log output:
```
root@f1a8dc97aafd:~# ./blueiris.sh
0044:fixme:service:scmdatabase_autostart_services Auto-start service L"BlueIris" failed to start: 2
002c:err:module:process_init L"C:\\windows\\system32\\bash.exe" not found
root@f1a8dc97aafd:~# 00c0:fixme:heap:RtlSetHeapInformation 00000000019A0000 0 000000000031E890 4 stub
00c0:fixme:nls:RtlGetThreadPreferredUILanguages 00000038, 000000000031E554, 000000000031E570 000000000031E550
00c0:fixme:nls:get_dummy_preferred_ui_language (0x38 000000000031E554 000000000031E570 000000000031E550) returning a dummy value (current locale)
0060:err:service:validate_context_handle Access denied - handle created with access 15, needed 2
00c0:fixme:shell:InitNetworkAddressControl stub
```

TonyBrobston commented:
I also saw someone say "I think you have to run the executable as administrator for it to allow “install as service”." on reddit. I really have no idea if this is necessary or possible: https://www.reddit.com/r/BlueIris/comments/g5ki6h/failed_to_install_blueiris_5_as_a_service/

TonyBrobston commented:
Here is the window where it asks for username/password for reference:
![image](https://user-images.githubusercontent.com/4724577/92138714-5994e000-edd4-11ea-9c1c-91ce1a6ad0e6.png)

leonowski commented:
I guess I should have been more clear on what this PR does.  I made it because of the problem of trying to auto-create the service through the Blue Iris GUI (it does not work).  So, these changes I put in *automatically* sets it up as a service upon installation.  It will then kill the container.  After starting the container again (with the persistent data store /root/prefix), the GUI launch will run in the 2 process mode (GUI + service).  If you check the options, it will have the "Run as Windows service" box checked

I put these in the notes, but there are new run options that need to be specified:

`--init` - to allow the internal script to kill PID 1
`-e TZ=yourtimezone` - to pass the TZ env var

In addition to this, you must mount `/root/prefix` as a persistent store somewhere on your system.  Also, privileged mode is still required during the install.

TonyBrobston commented:
@leonowski Is it possible for me to disabled the "start as a service" without opening the Blue Iris UI?

TonyBrobston commented:
Looks like maybe switching:
`"Start"=dword:00000002` to `"Start"=dword:00000003`?

in

```
[System\\CurrentControlSet\\Services\\BlueIris] 1599146726
#time=1d68206731a6b9c
"Description"="Video security software"
"DisplayName"="Blue Iris Service"
"ErrorControl"=dword:00000001
"ImagePath"="\"C:\\Program Files\\Blue Iris 5\\BlueIris.exe"
"ObjectName"=".\\root"
"PreshutdownTimeout"=dword:0002bf20
"Start"=dword:00000002
"Type"=dword:00000010
```

leonowski committed:
add mfc42, taskmgr seems to keep service alive

leonowski commented:
So, I checked in a couple of things into this strange branch.  It looks like mfc42 was another dependency for something in blueiris, so I added that.  I also noticed that the service does indeed run with this build, but it kept closing fairly quickly for some unknown reason.  To my surprise, it seemed like launching the windows Task Manager in the container kept the service running.  I will still have to explore this - wine is very strange and I'm new to it.  Actually, it seems like if I keep any windows process up, the service stays up (like a cmd prompt).  But, taskmgr seems to be a good thing to run since you can keep track of the blueiris.exe.

If this service stuff works out, we could have 2 modes when launching this container controlled by env vars:  A "headless" mode that just runs the service and whatever minimal components are needed and a admin mode that can launch the management GUI to do things like add cameras and such.

TonyBrobston commented:
@leonowski I'm going to pull the latest on that and run it for awhile.

I've been having Blue Iris randomly lock up about once a day. Once it's locked up I can't seem to do anything. A docker restart on the container brings it up just fine, but while it was locked up it isn't recording. It also happens to have the update screen up, may or may not be related. 
![image](https://user-images.githubusercontent.com/4724577/92248140-d5069800-ee8d-11ea-9a61-e8fd11f0967a.png)

I'll report back later, after running your latest updates for the day.

TonyBrobston commented:
I also noticed that the time/timezone is correct everywhere except the timestamp that is in the camera view. However when I used this solution before, it worked everywhere: https://github.com/jshridha/docker-blueiris/issues/17#issuecomment-685886361

Here is an example of the time/timezone. You can see that the time on my mac and the time on the ubuntu vnc thing match, but the time in the video window does not.
![image](https://user-images.githubusercontent.com/4724577/92250358-fc129900-ee90-11ea-89be-261df631782c.png)

Also, I haven't said this enough. Thank you for all the work you have been doing! It seems like you're very knowledgable and I really appreciate that you're diving deep into this. Let me know in what ways I can help, as I'll fill whatever gaps I can.

leonowski committed:
- shrink image by reducing number of RUN commands in Dockerfile
- Use ENTRYPOINT now with default supervisord-normal.conf.
  alernate conf can be use as CMD "-c supervisord-service.conf"
  to run a GUI-less and service only process.

leonowski commented:
This latest check-in for this branch seems to be working better.  I need to update the README still, but here are the important points:

- Persistent data store /root/prefix is required.  Everything else is in the ephemeral container.  So, make sure you set up a mount for /root/prefix so you can persist that data (it's the wine environment).

- Task Manager is launched by supervisord at startup. This has the benefit of bringing up the Wine Mono and Wine Gecko installation dialogs as soon as the container is first launched.  You no longer have to wait for the  blueiris.exe installer to download first.  Yay.

- After installation, uncheck the 2 checkboxes at the end of the wizard that allow you to read the blueiris readme and launch blueiris immediately.  You want to skip this because after the wizard is done, the blueiris.sh script will kill PID 1 to stop the container (you must then restart the container one more time or have originally set the container to `restart=always`.

- After it comes back up, the GUI should launch.  At this point, you can go ahead and set up the app as you like.  Add cameras, set up the web GUI interface, etc.  After you finish, destroy the container.  This will leave you with the persistent data.

- Now, here's the fun part:  You can run this as a docker container that runs the blueiris.exe as a service only.  X11, novnc, X11vnc, and flux box will not run.  It will run as a pretty tight container that only runs blueiris.exe and some supporting processes (supervisor, a windows dummy CMD window to keep the blueiris service running, and a process watching script that will force the container to die if blueiris.exe crashes).  This makes the container behave like a real service.  You can then work with blueiris through the UI3 interface (on port 81).  To run as a service, launch the container again using the existing persistent data and a command:  ` -c /etc/supervisor/conf.d/supervisord-service.conf` .  This will force the container to use the alternate supervisor config that only runs the service.

The downside to this is that if you want to go in and do things like add more cameras or manage the server, you will have to destroy the container again, and then relaunch without the ` -c /etc/supervisor/conf.d/supervisord-service.conf`.  Things should restart fairly quickly, but blueiris will be down for that short period of time between restarts.

Other interesting notes about this version:

- I removed the multiple uses of RUN in the Dockerfile in an attempt to shrink the file system layers.  It seems to have worked.  This image is still pretty huge though.  It might be good to have another image that just has minimal components when running in service-only mode.

- ENTRYPOINT (/usr/bin/supervisord)  is now used with a default CMD.  This allows us to change the launch options easily using CMD at run time.

- Finally, I'm still having problems with recording continuously on 64-bit. :(  Maybe it's the kernel I'm using.  I'm on 4.19.  I'm curious to know what you're using @TonyBrobston )

TonyBrobston commented:
@leonowski I'll give this last post a read when I have some time and update to the latest (probably tonight or tomorrow afternoon).

As for the kernel version, I googled and found a few commands:
```
server:~$ uname -r
4.15.0-91-generic
server:~$ uname -a
Linux server 4.15.0-91-generic #92-Ubuntu SMP Fri Feb 28 11:09:48 UTC 2020 x86_64 x86_64 x86_64 GNU/Linux
server:~$ cat /proc/version
Linux version 4.15.0-91-generic (buildd@lgw01-amd64-013) (gcc version 7.4.0 (Ubuntu 7.4.0-1ubuntu1~18.04.1)) #92-Ubuntu SMP Fri Feb 28 11:09:48 UTC 2020
```

TonyBrobston commented:
I had a spare minute so I updated things as you mentioned above. Also my timezone issue I described a few comments back seems to be gone.

I'll run things as a service for a day or two and report back.

Here's my docker-compose for reference:
```
 blue-iris:
    image: leonowski/docker-blueiris:run_as_service
    restart: unless-stopped
    privileged: true
    init: true
    environment:
    - RESOLUTION=1440x768x24
    - TZ=America/Chicago
    ports:
    - 8080:8080
    - 5900:5900
    - 81:81
    volumes:
    - /tmp/blue-iris:/root/prefix:rw
    command: -c /etc/supervisor/conf.d/supervisord-service.conf
```

leonowski committed:
- disable use of prgrep in check_process due to bug: https://bugs.launchpad.net/ubuntu/+source/procps/+bug/1874824
- tweak service startup again

TonyBrobston commented:
I spun up another container to test this setup: https://github.com/danecreekphotography/node-deepstackai-trigger/wiki/Integrating-with-BlueIris

In the process I noticed the time/timezone in the video window was wrong again. I did a quick `docker restart` on that container and the time/timezone immediately corrected itself.

TonyBrobston commented:
@leonowski I just now realized what you meant when you said "recording continuously". I thought you were referring to just hitting  this button:
![image](https://user-images.githubusercontent.com/4724577/92335392-22d0fb00-f05c-11ea-9944-fe792a457a42.png)

I'm messing around with setting up `node-deepstackai-trigger` and am having some issues when I select:
![image](https://user-images.githubusercontent.com/4724577/92335417-557af380-f05c-11ea-9673-c95336148e2f.png)
Basically the Blue Iris program stays running but the cameras just freeze.

leonowski commented:
I was able to determine the problem.  The problem happened when recording started and these options were turned on:

- Direct to disk recording
- Overlay

Direct to disk recording should allow the h264/h265 streams from my cameras to be saved directly to disk without re-encoding by blueiris.  I believe the problem happens because turning on overlay actually requires re-encoding since it is adding information to the original stream.  In any case, I don't need to use the overlay feature since my cameras can do this directly and this can save some CPU cycles on blueiris.  We should add this problem to the notes.

The interesting problem that arose from this is that blueiris would get in a state of continuous restarts if something bad happens with the camera settings.  To get out of this, you can disable the camera by changing the registry under:

HKEY_LOCAL_MACHINE\SOFTWARE\Perspective Software\Blue Iris\Cameras\(your_camera)name\

There is a value for "Enabled" set to 1.  If you set it to 0 and restart the container, you can get out of this state.  You can change the .reg files in a text editor directly (in the wine prefix /root/prefix).  

This got me thinking - it might be good to have a script or some type of "safe mode" that disables all cameras.  This will allow you to at least get into the Blue Iris GUI to fix some things.

TonyBrobston commented:
@leonowski Have you tried using this Blue Iris docker image with `node-deepstackai-trigger`? I'm trying to trigger my camera. I am able to get `node-deepstackai-trigger` to call the correct Blue Iris endpoint. I seem to be having some sort of networking issue. The Blue Iris container/service/network name is `blue-iris`. Here's what the logs say: 
```
[Web request] /aiinput/person.jpg: Calling trigger uri http://blue-iris:81/admin?trigger&camera=house_east&user=[redacted]&pw=[redacted]
```

When I run the same thing from a web browser on my development machine substituting `blue-iris` with my server's IP, it triggers the camera just fine:
```
http://[my server's IP here]:81/admin?trigger&camera=house_east&user=[redacted]&pw=[redacted]
```
However if I do the same thing from `triggers.json` it will show:
```
[Web request] /aiinput/person1.jpg: Calling trigger uri http://[my server's IP here]:81/admin?trigger&camera=house_east&user=[redacted]&pw=[redacted]
```
but still it won't trigger the camera.

leonowski commented:
Since you're exposing port 81 on the host, you can use the host ip instead of the container name and it should work.

TonyBrobston commented:
@leonowski I tried that, unfortunately no luck 😢 .

leonowski commented:
As a test, you can launch the container and override the entrypoint and user like this:

`docker run -it --entrypoint=/bin/bash --user=root --rm danecreekphotography/node-deepstackai-trigger`

Once inside the shell, do a `apt update` followed by `apt install curl`.  Then, you can try to use curl from inside this test container.  Maybe it can help you figure out the networking issue.  You can install other tools too to help you figure out things from inside the container.

TonyBrobston commented:
@leonowski It looks like it's an issue with the security around Blue Iris' web server. I just pulled down the `danecreekphotography/node-deepstackai-trigger` repo and added a `console.log` to the `Promise.all` to print the return of this call: 

https://github.com/danecreekphotography/node-deepstackai-trigger/blob/5241a18739254f1f2f4f2012cc67cb55eef38fa6/src/handlers/webRequest/WebRequestHandler.ts#L42

Looks like it returns the entire Blue Iris login page. So, seems like I have something configured incorrectly in Blue Iris to allow this to work.

TonyBrobston commented:
Seems like the issue is that docker is assigning a `172` address and I think Blue Iris isn't allowing `172` addresses. I'm not quite sure what to do in this case. I don't really know the difference between `192` and `172` addresses...

leonowski commented:
The docker default subnet for container network is 172.17. 0.0/16.  So, you can get any IP from that for containers.  That is in the private range like 192 addresses.  It shouldn't affect your ability to reach the IP you can reach blueiris at.  I just tested this and blueiris's web server works fine inside the docker container from other containers.  You can test this by trying to curl or netcat the private IP address and port from the host or other containers.

How are you webserver settings set up?  `Use Secure session keys and login` should be turned off.  You may want to turn off auto banning as well.  Also, set authentication to be required from all connections.

<img width="496" alt="2020-09-07 11_28_34-blueiris - Remote Desktop Connection" src="https://user-images.githubusercontent.com/1335759/92411686-beff0e80-f0fd-11ea-9199-3cd6262aa0ab.png">

TonyBrobston commented:
@leonowski You're my savior man. That worked. Thank you!

leonowski committed:
- fix ZIP command to overwrite existing files.  Useful for upgrades.
- update README
- Add dot character to PATHEXT in registry.  This allows wine to run scripts in linux.  Useful for blue iris alerts.

leonowski commented:
This last update adds the ability to run Linux scripts from inside the WIndows/Wine environment.  This is good for configuring alerts in Blue Iris.  Got it from this:

https://wiki.winehq.org/FAQ#How_do_I_launch_native_applications_from_a_Windows_application.3F

That way, you don't have to do any fancy windows stuff to trigger something.  Just call a trusty linux script to do something (like curl another web service).

TonyBrobston commented:
@leonowski I grabbed the latest again and am running it.

I've noticed a lot of: 
```
00f4:fixme:variant:get_date_format ignoring flags 400
``` 
in the logs, any idea what that is?

I've also been seeing a lot of:
![image](https://user-images.githubusercontent.com/4724577/92513886-a1ea3e80-f1d6-11ea-9fb9-7b519ea2b11b.png)
in the web UI, I've had to switch from the HTML5 player to JavaScript to make it usable; otherwise it refreshes the UI every few seconds.

leonowski commented:
I don't know what the get_date_format warning message is for.  I have not looked into it yet.  I see the same thing.

I have the same problem in the Web UI as well and have to switch to the javascript player too.  I then tested it against a real Windows installation and I'm running into the same issue.  It seems to be a problem with Blue Iris.  Haven't found a solution yet.

TonyBrobston commented:
@leonowski With a bit of digging, looks like others are having he same issue as of recently. I wonder if it'd be worth eventually working towards tagging docker images based on blue iris versions? That way we could swap a tag to switch easily?

https://ipcamtalk.com/threads/video-delay-has-exceeded-60-seconds.50796/

leonowski commented:
The issue was actually outside of Blue Iris and with the UI3 package.  Extracting the latest version to the UI directory in your wine prefix will fix this:  https://github.com/bp2008/ui3/releases/tag/118

The container is not set up to keep track of versions and Blue Iris doesn't seem to keep a storage area for historical releases.  They just have their main download.  We could change the startup script to use an ENV VAR for the download URL if you have your own file that you want to use.

TonyBrobston commented:
@leonowski In an attempt to upgrade versions I deleted `BlueIris.exe` from here:
https://github.com/leonowski/docker-blueiris/blob/bb32638307ca4426c65a44522ae968dd6e9c757d/blueiris.sh#L3

It seems to download using: 
https://github.com/leonowski/docker-blueiris/blob/bb32638307ca4426c65a44522ae968dd6e9c757d/blueiris.sh#L19
Then at some point the container dies and is restarted, then it downloads again, and then the same over and over and over until I stop it.

I also get this error message in the UI:
![image](https://user-images.githubusercontent.com/4724577/92529848-e08cf280-f1f0-11ea-8d77-453903366b51.png)

leonowski committed:
- reduce sleep command times
- set normal run to not auto restart blueiris.  this will allow for updates through the GUI.

leonowski commented:
Hi @TonyBrobston 

Sorry about that.  I was wrong about the upgrade process.  I had tested it using the same exe install files, but not with an actual newer build of the installer exe.  It looks like this was the wrong way to go.  Do you still have access to the deleted blueiris.exe?  If not, I can post them somewhere for you to get.  I had renamed mine instead of deleting them.  

I have changed the supervisor config to not restart blueiris.sh.  That will allow you to continue running in the X11 desktop environment after the main blue iris processes close.  So, you can do the update using the update button in the Blue Iris settings.  I realize this doesn't help you now after deleting the exe - so, let me know if you need it.  Once you get the blue iris GUI up, run the process to update and it should close and restart blueiris for you.

TonyBrobston commented:
@leonowski I ended up just manually reconfiguring my Blue Iris. Since I've been in there a lot it only took me like 20 minutes.

leonowski committed:
- added menu items to fluxbox right click menu to manage blueiris
- added a script to update to latest UI3 from github

leonowski commented:
Sorry for such a long PR with so many comments and changes.  I think this is good to go if you guys think so too.

I changed the way the service behaves in the default mode.  The container will not stop when blueiris stops in the normal mode.  It will stay up so you can do things like updates or manage the wine environment.  It will still stop if you switch to the service mode.

I added a menu for Blue Iris stuff in the right click menu of Fluxbox.  You should see it under a main "Blue Iris" menu item in fluxbox.  Just right click on the desktop to see it.  I included:

- A way to launch the Blue Iris GUI.
- Ways to stop and start the Blue Iris windows service.
- A script to get the latest UI3 and unzip it to WWW

If you guys think this is good, I think it should go in the latest dev.

leonowski commented:
Forgot to mention:  Added the curl and jq packages.  Helpful for running scripts in container and will probably be useful for integration with the AI triggers.

Also, I removed the cifs-utils package.  I don't really think it's needed.

leonowski commented:
this PR should

close #17 
close #28

TonyBrobston commented:
@leonowski I'll pull the latest of this branch and run it.

TonyBrobston commented:
@jshridha I think we should consider merging this to `dev` soon. I've been running the updates and things have been good. I do have a small concern that it may eventually randomly crash; so it would be good to have it long running without stopping/starting the container.

I'm also currently using the node ai tool and have everything running in docker. It has been working great, compared to just plain old motion detection. I've been working on fine tuning my triggers. Unfortunately my neighbor put a t-shirt over a mail box (long story) and deepstack thinks there's a 85-97% chance it's a person 🤦  (depending on the lighting and different things). I'll likely start to play around with the mqtt part and tie some automations into Home Assistant.

Here's an example docker-compose (I realize this is a bit out of scope):
```
version: '3.8'

services:
  home-assistant:
    image: homeassistant/home-assistant
    restart: unless-stopped
    network_mode: host
    volumes:
    - ~/home-assistant:/config
    devices:
    - /dev/ttyACM1:/dev/ttyACM0

  blue-iris:
    image: leonowski/docker-blueiris:run_as_service
    restart: unless-stopped
    privileged: true
    init: true
    environment:
    - RESOLUTION=1440x768x24
    - TZ=America/Chicago
    ports:
    - 8080:8080
    - 5900:5900
    - 81:81
    volumes:
    - /SURVEILLANCE/blue-iris:/root/prefix:rw
    - aiinput:/root/prefix/drive_c/BlueIris/aiinput
    command: -c /etc/supervisor/conf.d/supervisord-service.conf

  trigger:
    image: danecreekphotography/node-deepstackai-trigger:latest
    restart: unless-stopped
    environment:
      - TZ=America/Chicago
    secrets:
      - triggers
      - settings
    ports:
      - 4242:4242
    volumes:
    - aiinput:/aiinput

  deepstack-ai:
    image: deepquestai/deepstack:latest
    restart: unless-stopped
    environment:
      - VISION-DETECTION=True
    volumes:
      - /SURVEILLANCE/deepstack/datastore:/datastore

secrets:
  settings:
    file: ./settings.json
  triggers:
    file: ./triggers.json

volumes:
  aiinput:
```

luck-y13 commented:
I've been trying to run leonowski/docker-blueiris:run_as_service on my unraid server with no success.  The blueiris.sh fails every time.  

Here is the docker command: 
docker run -d --name='BlueIris5' --net='bridge' --privileged=true -e TZ="America/Denver" -e HOST_OS="Unraid" -e 'TZ'='America/Denver' -e 'RESOLUTION'='1440x768x24' -p '8083:8080/tcp' -p '84:80/tcp' -p '5900:5900/tcp' -v '/mnt/user/appdata/blueiris5/bi/':'/root/prefix':'rw' --init 'leonowski/docker-blueiris:run_as_service' 

Here is the log
`
130150K .......... 100% 83.7M=2m40s

2020-09-10 16:21:34 (811 KB/s) - ‘blueiris.exe’ saved [133284136/133284136]

------------------------------------------------------
Running Wine/winetricks as root is highly discouraged. See https://wiki.winehq.org/FAQ#Should_I_run_Wine_as_root.3F
------------------------------------------------------
Executing mkdir -p /root
0024:err:environ:run_wineboot boot event wait timed out
`


![image](https://user-images.githubusercontent.com/69563599/92818364-dc91d980-f384-11ea-8eb4-f9b48eae3432.png)

leonowski commented:
Hi @luck-y13 

After it starts, are you able to go to the web interface on port 8083?  The blueiris script runs and it looks like something downloads, but there are required dialogs to jump through the GUI.  The first thing you should see is an attempt to install wine-mono and wine-gecko.  Eventually, the blue iris installer GUI should appear.

Do you get anything in the GUI?

luck-y13 commented:
I get this Wine popup saying that /root/prefix is being updated.  Then in the log it says that the wine timed out.

![image](https://user-images.githubusercontent.com/69563599/92823580-b111ed80-f38a-11ea-8ac6-1f97b1e66589.png)

leonowski commented:
I believe the privileged flag should be:  `--privileged` and not `--privileged=true`

luck-y13 commented:
I removed my data folder and removed the image leonowski/docker-blueiris:run_as_service and tried to start it again, but this time using leonowski/docker-blueiris:latest.  This time it had the pop ups for mono and gecko which installed fine and it prompted to install BI which also installed ok.  

But now the log has 
2020-09-10 17:47:24,120 INFO exited: blueiris (exit status 53; not expected)
2020-09-10 17:47:27,124 INFO spawned: 'blueiris' with pid 939
002b:fixme:actctx:parse_depend_manifests Could not find dependent assembly L"Microsoft.Windows.Common-Controls" (6.0.0.0)
002b:err:module:import_dll Library VCRUNTIME140_1.dll (which is needed by L"C:\\Program Files\\Blue Iris 5\\BlueIris.exe") not found
2020-09-10 17:47:28,170 INFO success: blueiris entered RUNNING state, process has stayed up for > than 1 seconds (startsecs)
002b:err:module:import_dll Library mfc140u.dll (which is needed by L"C:\\Program Files\\Blue Iris 5\\BlueIris.exe") not found
002b:err:module:import_dll Library FTD2XX.dll (which is needed by L"C:\\Program Files\\Blue Iris 5\\SeaMAX.dll") not found
002b:err:module:import_dll Library SeaMAX.dll (which is needed by L"C:\\Program Files\\Blue Iris 5\\BlueIris.exe") not found
002b:err:module:LdrInitializeThunk Importing dlls for L"C:\\Program Files\\Blue Iris 5\\BlueIris.exe" failed, status c0000135

luck-y13 commented:
> I believe the privileged flag should be: `--privileged` and not `--privileged=true`

I changed it to `--privileged` and no difference.

leonowski commented:
I just tried this in unraid as well (I haven't really tried it on there, but I'm actually a very early user of Unraid so I had one running so I could try it).

I had to make sure the persistent directory was clear for the first launch and then use `--privileged` as an extra parameter in the Advanced view.  Then it worked.  The final output of the command according to the Unraid GUI was:

root@localhost:# /usr/local/emhttp/plugins/dynamix.docker.manager/scripts/docker run -d --name='blueiris' --net='bridge' -e TZ="America/Los_Angeles" -e HOST_OS="Unraid" -e 'TZ'='America/Los Angeles' -p '8080:8080/tcp' -p '81:81/tcp' -p '5901:5900/tcp' -v '/mnt/cache/blueiris-test':'/root/prefix':'rw' --init --privileged 'leonowski/docker-blueiris:run_as_service' 
028e511b28023788ba1ce6971d7426469efb01830dca3dd528d060b21164713e

The command finished successfully!



![image](https://user-images.githubusercontent.com/1335759/92836042-d0137e00-f390-11ea-8085-1c688331d3bd.png)

if it still fails, can you post the entire output of `docker logs` for the container?

TonyBrobston commented:
I'm having a problem that I think is probably unrelated to this branch. However I'm curious to know if anyone else experiences this issue. After the video container has been running for a long time I get weird speeding up and slowing down of frames on all of my cameras in Blue Iris (I think I had this same issue on Unifi Video, after that docker container was long running). If I restart the container the problem seems to go away. I'll attach a video showing an example (it wouldn't let me upload a `.mp4`, so I put it in a folder and zipped it, since it allows zip).
[video.zip](https://github.com/jshridha/docker-blueiris/files/5210325/video.zip)

luck-y13 commented:
You got it to work on Unraid which gives me hope, but I still can't get it to install with  your run_as_service branch.  I've deleted my image and re-downloaded it multiple times.  My docker run is the same as yours except for the TZ.  I go to 8083 and see it saying the wine configuration is being updated, but it just hangs and never brings up the gecko or mono installer.

Here is the log.  You'll see after it downloads the blueiris.exe, it tries to launch it with wine, but I guess since it never installed gecko or mono it just hangs.

[blue-iris-docker-log.txt](https://github.com/jshridha/docker-blueiris/files/5211615/blue-iris-docker-log.txt)

leonowski commented:
Is there anything in your wine cache directory in the container?  ` ~/.cache/wine`

leonowski committed:
add wine gecko and mono msis to package

leonowski commented:
After it finishes building on dockerhub, give this new one a try.  I put in gecko and mono msis into the package.

leonowski committed:
correct way to prebake mono and gecko msis.  change mv to cp in blueiris.sh to avoid errors.

leonowski commented:
ok, this newer version seems better at prebaking mono and gecko.  hopefully, this fixes the issue.

leonowski committed:
removed extra wgets

johnny2678 commented:
First off, thanks for this.

I have been trying to get your `run_as_service` branch for days. Was trying to figure this out on my own, but keep bumping into the same problem.

running on Unraid with this command:
```
docker run -d --name='blueiris' --net='bridge' -e TZ="America/New_York" -e HOST_OS="Unraid" -e 'RESOLUTION'='1440x900x24' -p '8123:8080/tcp' -p '8987:81/tcp' -p '5903:5900/tcp' -v '/mnt/user/appdata/bi5-leonowski/':'/root/prefix':'rw'  --log-opt max-size=5m --log-opt max-file=2 --init --privileged 'jh/bi5:leonowski' 
```

have tried creating the local share and chown-ing to nobody:users and also just letting docker create the share specified in the run cmd, but the container fails with multiple messages like it can't find the `/root/prefix`

such as:
```
wine: '/root/prefix' is not owned by you
2020-09-12 13:08:43,293 INFO exited: dummy-taskmgr (exit status 1; not expected)
Warning: Failed to open file(/usr/share/fluxbox/nls/C.UTF-8/fluxbox.cat)
```
...
```
Executing wine regedit /S C:\windows\Temp\override-dll.reg
regedit: The file 'C:\windows\Temp\override-dll.reg' was not found.
Executing wine64 regedit /S C:\windows\Temp\override-dll.reg
regedit: The file 'C:\windows\Temp\override-dll.reg' was not found.
```
...
```
------------------------------------------------------
Note: command wine vc_redist.x86.exe /q returned status 2. Aborting.
------------------------------------------------------
0170:err:module:process_init L"C:\\windows\\system32\\blueiris.exe" not found
rm: cannot remove 'blueiris.exe': No such file or directory
unzip:  cannot find or open /root/prefix/drive_c/Program Files/Blue Iris 5/ui3.zip, /root/prefix/drive_c/Program Files/Blue Iris 5/ui3.zip.zip or /root/prefix/drive_c/Program Files/Blue Iris 5/ui3.zip.ZIP.
00b0:fixme:ntdll:NtQuerySystemInformation info_class SYSTEM_CACHE_INFORMATION
00b0:fixme:ntdll:NtQuerySystemInformation info_class SYSTEM_CACHE_INFORMATION
00b0:fixme:ntdll:NtQuerySystemInformation info_class SYSTEM_CACHE_INFORMATION
unzip:  cannot find or open /root/prefix/drive_c/Program Files/Blue Iris 5/ui3.zip, /root/prefix/drive_c/Program Files/Blue Iris 5/ui3.zip.zip or /root/prefix/drive_c/Program Files/Blue Iris 5/ui3.zip.ZIP.
```

full pastebin:
https://pastebin.com/ym5Z4S7C


any ideas?

johnny2678 commented:
Just realized you have this on docker hub.  Tried again with that image in case there was something wrong with my local build from your git branch.

Current docker run cmd:
```
docker run -d --name='blueiris' --net='bridge' -e TZ="America/New_York" -e HOST_OS="Unraid" -e 'RESOLUTION'='1440x900x24' -p '8123:8080/tcp' -p '8987:81/tcp' -p '5903:5900/tcp' -v '/mnt/user/appdata/bi5-leonowski/':'/root/prefix':'rw'  --log-opt max-size=5m --log-opt max-file=2 --init --privileged 'leonowski/docker-blueiris:run_as_service'
```

almost the same results unfortunately:

This time I didn't VNC into the container until the logs stopped scrolling.  Was greeted with this screen:
![Screen Shot 2020-09-12 at 2 10 54 PM](https://user-images.githubusercontent.com/8219743/93002108-4edbf880-f502-11ea-9f75-d4547479ff66.png)

pastebin: https://pastebin.com/YcdULADZ

leonowski commented:
It does look like some permissions issue.  I tried to run this on a mac with Docker Desktop and I'm running in to the same problem.  It seems like wine tries to  link serial ports of the host machine in ~./wine/dosdevices at some point.  I believe this fails and so everything else fails.  On the mac, I was able to get around this by using a named volume just for the prefix directory like this:

`somevolume:/root/prefix` 

As opposed to mapping a volume to a path like this:

`$(pwd)/some_place:/root/prefix`

As a test, can you try creating a new volume and using that instead of a path?  I'm not sure how to do that through the unraid GUI, but you can do it through the command line like above.  If you specify a volume name that doesn't exists yet, it should create it (or, you can make one before you start with `docker volume create some_volume_name`

johnny2678 commented:
ok, trying again with this after creating a volume:
```
docker run -d --name='blueiris2' --net='bridge' -e TZ="America/New_York" -e HOST_OS="Unraid" -e 'RESOLUTION'='1440x900x24' -p '8123:8080/tcp' -p '8987:81/tcp' -p '5903:5900/tcp' --log-opt max-size=5m --log-opt max-file=2 --init --privileged --mount source=bi5,target=/root/system 'leonowski/docker-blueiris:run_as_service' 
```

This does allow me to proceed to the BI install.  It installs and then shuts down the container after completion as planned.

However upon restart and removing `--privileged` I get a bunch of explorer errors

```
012c:err:winediag:nodrv_CreateWindow Application tried to create a window, but no driver could be loaded.
012c:err:winediag:nodrv_CreateWindow The explorer process failed to start.
012c:err:ole:apartment_createwindowifneeded CreateWindow failed with error 0
2020-09-12 17:31:31,856 INFO spawned: 'dummy-taskmgr' with pid 934
012c:err:ole:apartment_createwindowifneeded CreateWindow failed with error 0
012c:err:ole:apartment_createwindowifneeded CreateWindow failed with error 0

```

also, is NoVNC working for you guys? long list of errors in the pastebin starting with this:
```
WebSocket server settings:
  - Listen on :8080
  - Web server. Web root: /root/novnc
  - No SSL/TLS support (no cert file)
  - proxying from :8080 to localhost:5900
Failed to read: session.ignoreBorder
Setting default value
Failed to read: session.forcePseudoTransparenc
```

Not trying to create work for you if you've got it working.  Was just trying to move my install to docker after accidentally borking my win10 VM ;)

full pastebin: https://pastebin.com/krd2R4bz

leonowski commented:
No worries.  This is all interesting to me.  It may take some time to figure out as I'm relatively "new" to wine.  I'm sure it has to do with running things as root.  I'm looking into not having to do that now.

leonowski commented:
@johnny2678 

I just noticed in the run you pasted this volume:

`--mount source=bi5,target=/root/system`

Did you mean the target to be `/root/prefix` instead?

johnny2678 commented:
OMG... 🥴  that's embarrassing.

Still no NoVNC
![image](https://user-images.githubusercontent.com/8219743/93006852-64b3e280-f52f-11ea-89cb-49d28aab53a7.png)

But over VNC the install completed and the console came up after the container restarted.  Need to spend some more time with this but I think I've taken enough of your Saturday ;) sorry.

leonowski committed:
- removed privilged requirement
- run container as user wineuser with uid 1000

leonowski commented:
ok, i think this should work for you @johnny2678 .  You can run it without privileged mode now.  But, you must create the prefix path first on the host machien and set it to 777 so the container can write to it.  Or, if you used a named volume, that isn't required.

leonowski commented:
Another note:

Now that this doesn't run as root, the default port for the blueiris web server can not use port 81.  So, in the GUI, you must set it to something higher than 1024.  You can still map it to anything you want through docker though like `-p 81:1081`

luck-y13 commented:
Thanks for you working with us on this @leonowski!  I really appreciate your time.  I tried the home/wineuser update and dropped the --privileged flag.  I am using a local path to a directory that I chown 777 on.  I even created a subdirectory within that folder that I also chown 777'd.  

`root@localhost:# /usr/local/emhttp/plugins/dynamix.docker.manager/scripts/docker run -d --name='BlueIris5-run-as-service' --net='bridge' -e TZ="America/Denver" -e HOST_OS="Unraid" -p '8083:8080/tcp' -p '84:80/tcp' -p '5901:5900/tcp' -v '/mnt/user/blueiris_install/data/':'/home/wineuser/prefix':'rw' --init --log-opt max-size=5m --log-opt max-file=2 'leonowski/docker-blueiris:run_as_service' `

Now I get no popups on the vnc, not even taskmanager and the log says

`2020-09-13 13:10:15,787 INFO spawned: 'dummy-taskmgr' with pid 49
wine: '/home/wineuser/prefix' is not owned by you
2020-09-13 13:10:15,795 INFO exited: dummy-taskmgr (exit status 1; not expected)
--2020-09-13 13:10:17-- https://blueirissoftware.com/blueiris.exe
Resolving blueirissoftware.com (blueirissoftware.com)... 64.20.60.68
Connecting to blueirissoftware.com (blueirissoftware.com)|64.20.60.68|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 133284136 (127M) [application/octet-stream]
Saving to: ‘blueiris.exe’
`

Here is what the permissions show inside the docker container:
![image](https://user-images.githubusercontent.com/69563599/93026579-346a5380-f5c4-11ea-885a-25c2762b0b34.png)

Here is what the permissions show outside the container.  The data folder which is mapped to prefix was created and owned by root:root, but it gets changed to 1000:1000 by the dockerfile.  I'm guessing this is because the user and group wineuser:wineuser doesn't exist outside the docker container context?

![image](https://user-images.githubusercontent.com/69563599/93026790-ddfe1480-f5c5-11ea-97da-323273e77bf2.png)

[blue-iris-docker-log-9-13-2020.txt](https://github.com/jshridha/docker-blueiris/files/5214991/blue-iris-docker-log-9-13-2020.txt)

leonowski committed:
- clean prefix on first run
- removed taskmgr task as it blocks clean first run
- use winetricks annihilate to create clean prefix
- added notes to indicate that privileged mode is still needed on first run

leonowski commented:
It seems like a big problem was with the way wine was being initialized.  winetricks ran during the build and I think that throws a wrench in things.  This new image now does no wine operations during build.

So, on first launch now, it will create the fresh wine prefix environment for you using the `winetricks annihilate` command to start off.  On unraid, you should still create the folder first if your using a volume mapping.  You should not need to set 777, but you can chown using the UID like this:

`chown -R 1000:1000 /your/path`

Invoke it through docker:  `-v /your_path:/home/wineuser/prefix`

The consequence of this is the initial startup is now longer as it will download vc runtime 2019, mfc42, corefonts, and wininet.  When it's all done, it should eventually give you the Blue Iris installer.  

I had to remove the taskmanager task as it blocked other components from installing on the first run.  If you need to run it to take a look at things, you can still do it through the shell.  I add it as a right click menu item in fluxbox later.

Hopefully, this works.  I tested it on my unraid.  I forgot to mention:  We still need privileged mode on the first run.  I put that back in the readme.

TonyBrobston commented:
@leonowski I pulled the latest, deleted my container, and ran the latest. My setup is now broken.

When I remove the command that tells it to run as a service, things start up, it downloads Blue Iris even know I already have it installed, then does a few other things. Eventually I get a Blue Iris install screen in the VNC. I walk through that, it asks me for my License Key even know it was previously configured, I enter it and hit "Next", then Blue Iris stops running for some reason.

I was looking through the updates to the readme and I think the issue might be related to `wineuser`, I haven't ran any `chmod`'s. I'll give that a shot.

TonyBrobston commented:
I tried running a `chmod 777` from my host, on my volume mounted directory. Unfortunately I get the same result as above, it doesn't seem to find my Blue Iris install and configuration.

For now I cloned down you repo and then checked out a specific commit, built that, and am running on it just fine.
https://github.com/leonowski/docker-blueiris/commit/3ae19b7c0f2e85044f478b0fd44cbbb5e24bfb18

Any idea what I need to do to run on the latest?

leonowski commented:
Hi @TonyBrobston 

The big change is this container no longer runs as root.  The problem is, you ran the old version which ran as root.  Wine does not allow you to use the wine prefix unless the user owns the prefix.

The other big change is the prefix is now located at /home/wineuser/prefix

So, to fix the problem, first change the owner of your persistent store on the host to ID 1000 like this:

`chown -R 1000:1000 /path/to/your/data`

Then, change your docker-compose yml  to point from the old `/root/prefix` path to `/home/wineuser/prefix` path

leonowski committed:
fix owner of prefix dir if using named volume mapping
remove winetricks annihilate command as it is no longer needed

TonyBrobston commented:
@leonowski That fixed the issue.

I did have one other issue I struggled with for a bit. When using a volume like in my docker-compose above:
```
volumes:
  aiinput:
```
`aiinput` got created by `root` user at some point. So I think `wineuser` was not able to write to that folder as a result, breaking my ai trigger stuff. I was able to `chown -R 1000:1000 /aiinput` (from inside one of the container, I had to `exec` into it with something like `docker exec -it --user root <container id> /bin/bash` to be root user) which fixed the issue. I'm not sure if this folder gets created by `root` by default (through docker-compose) or if it was created by `root` because I previously ran as `root` when using Blue Iris. Since `aiinput` is essentially mapped from one container (`blue-iris`) to the other (`trigger`), that `aiinput` folder really doesn't end up on the host machine, which made this whole situation confusing.

TonyBrobston commented:
@jshridha Any update on merging this to dev? I've been working off of this branch for awhile now and things have been great.

jshridha commented:
@leonowski Are you finding this to be stable also? If so I'll merge.

Viksie commented:
From working 32 bit jshridha/blueiris and danecreekphotography/node-deepstackai-trigger images, to being stuck on a 64 bit broken setup...

For reference, my portainer (docker-compose) being as follows:

```
    blueiris:
        image: leonowski/docker-blueiris:run_as_service
        privileged: true
??      init: true			<- Unsupported config option for blueiris service: 'init'
        container_name: blueiris
        environment:
          - PUID=1000
          - PGID=1000       
          - RESOLUTION=1920x900x24
          - TZ=Australia/Sydney
        volumes:
# chown -R 1000:1000 /cam **
          - /cam:/home/wineuser/prefix:rw
          - /cam/tmp:/tmp
        ports:
          - 8055:8080
          - 5900:5900
# due to wineuser <- port > 1024
          - 8056:81
#       command: -c /etc/supervisor/conf.d/supervisord-service.conf
        logging:
          options:
            max-size: "5m"
            max-file: "2"
        restart: unless-stopped
```

I've started numerous times from scratch, yet lady luck is not smiling upon me. My burning questions being:

1. The required persistent data store is thus now:

	- /home/wineuser/prefix

2. What is the su or root password as to elevate wineuser(1000) ?

3. Is the config option --init still required with the initial setup ?

4. I can't get a web connection to host:81 although the BlueIris webserver is enabled and set to port 81 (172.0.0.5:81) of which the port is exposed. Under Web Server > Advanced > "Use secure session keys and login page" and "Auto-ban IP after attempts" are both unchecked. Running the script "get_latest_ui3.sh" brings no progress. 

```
$netstat -tulpan

Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name
tcp        0      0 0.0.0.0:5900            0.0.0.0:*               LISTEN      228/x11vnc
tcp        0      0 0.0.0.0:8080            0.0.0.0:*               LISTEN      24/python
tcp        0      0 127.0.0.1:5900          127.0.0.1:60604         ESTABLISHED 228/x11vnc
tcp        0      0 127.0.0.1:60604         127.0.0.1:5900          ESTABLISHED 229/python
tcp        0      0 172.17.0.4:8080         172.17.0.1:59500        ESTABLISHED 229/python
tcp6       0      0 :::5900                 :::*                    LISTEN      228/x11vnc
```

What am I missing to get the webserver going ... ?

5. The "playback timelime" and clip summaries (thumbnails) are displaying UTC+0 though only the cameras' overlays are displaying the localtime correctly. Surely it would be preferable to have the overlays off, as to save on CPU cycles, yet then the clips' and timeline reflecting the recorded time would be a necessity....

TonyBrobston commented:
@Viksie For reference, here is my latest docker-compose. This has been working great for me for a few weeks:
```
version: '3.8'

services:
  blue-iris:
    image: leonowski/docker-blueiris:run_as_service
    restart: unless-stopped
    # privileged: true
    init: true
    environment:
    - RESOLUTION=1440x768x24
    - TZ=America/Chicago
    ports:
    # - 8082:8080
    - 8081:8081
    volumes:
    - /SURVEILLANCE/blue-iris:/home/wineuser/prefix:rw
    - aiinput:/home/wineuser/prefix/drive_c/BlueIris/aiinput
    command: -c /etc/supervisor/conf.d/supervisord-service.conf

  trigger:
    image: danecreekphotography/node-deepstackai-trigger
    restart: unless-stopped
    environment:
    - TZ=America/Chicago
    secrets:
    - triggers
    - settings
    ports:
    - 4242:4242
    volumes:
    - aiinput:/aiinput

  deepstack-ai:
    image: deepquestai/deepstack
    restart: unless-stopped
    environment:
    - VISION-DETECTION=True
    volumes:
    - /SURVEILLANCE/deepstack/datastore:/datastore

secrets:
  settings:
    file: ./settings.json
  triggers:
    file: ./triggers.json

volumes:
  aiinput:
```

I ran into issues with using port 81 inside and outside the container, so I switched that to 8081 and also had to change that setting in Blue Iris' webserver. Also I believe timezone has been working for me, but I can double check.

Here are some screenshots of my webserver settings:
![image](https://user-images.githubusercontent.com/4724577/94850522-a926f580-03ec-11eb-9a2d-94bbd39cc31b.png)

![image](https://user-images.githubusercontent.com/4724577/94850556-b348f400-03ec-11eb-98dc-5ecb4f67331c.png)

Hopefully seeing a working docker-compose will clear up some of your issues. @leonowski may be able to weigh in more on some of your questions. I'm not using portainer, so I'm no help there.

Viksie commented:
Thanks @TonyBrobston. Re-reading the entire thread also helped getting it to work again.

However, I'm still seeing a difference in the time displayed within the console and UI3 interface. Clips in the console is displayed UTC+0 whereas clips in UI3 is displayed with localtime.

And even worse ... upon a restart of BI, ALL clips are being deleted, be they .bvr or .mp4, "Combine or cut video each" disabled or not. Once Bi has started up, the deleted clips are still reflected by empty placeholders in the clips section, until cleared by a DB repair.

I've noticed that the files (or entire folder) are being deleted just before the splash-screen of BI is displayed on startup.  Which has me ponder if it's not related to a script ...

TonyBrobston commented:
@Viksie I did a `docker restart` on my container and I still have video files from weeks ago and am still able to play them. I also went into the Blue Iris program and shut it down, I had issues using the menu to start Blue Iris (right click > Applications > I  tried the first two options I believe). I then did another restart on the docker container and all my video files are still there and playable. I also did a Blue Iris software update, which worked perfectly.

I'm not sure I'll be much help on the rest. Sorry 🤷 .

Viksie commented:
It actually clears the entire folder. Be it a container restart or the app itself from the console. 

Even when I created a new folder "aiclips", it does the same on this new folder. I don't see anything in the logs and is flabbergasted by the "behavior".

One thing it is, is being consistent on my system. I recall reading a thread may moons ago when some else had this issue and I scoffed at the notion ...

luck-y13 commented:
I'm still having trouble with Bl getting installed on my volume in unraid.  I've made all the changes suggested and created a volume that I chown 1000:1000.  The c_drive in the prefix ends up with a windows folder, but it doesn't create the user folders when the installer for BlueIris downloads and tries to run.  I was wondering if it might be the access type for the path in docker.  I have tried rw, rw-slave, rw-shared.  Doesn't seem to make a difference.  In the end the I still get 
![image](https://user-images.githubusercontent.com/69563599/95246428-b586c600-07d1-11eb-80b3-35cbc843cb4d.png)

Log

`130100K .......... .......... .......... .......... .......... 99% 1.02M 0s
130150K .......... .......... .......... .......... .......... 99% 37.4M 0s
130200K 100% 30.8M=92s

2020-10-05 09:56:48 (1.38 MB/s) - ‘blueiris.exe’ saved [133325776/133325776]

002c:err:shell:SHGetFolderPathAndSubDirW Failed to create directory L"C:\\users\\wineuser\\Application Data".

002c:err:shell:SHGetFolderPathAndSubDirW Failed to create directory L"C:\\users\\wineuser".

002c:err:shell:SHGetFolderPathAndSubDirW Failed to create directory L"C:\\users\\wineuser\\Local Settings\\Application Data".

0024:fixme:actctx:parse_depend_manifests Could not find dependent assembly L"Microsoft.Windows.Common-Controls" (6.0.0.0)
0040:err:shell:SHGetFolderPathAndSubDirW Failed to create directory L"C:\\users\\wineuser\\My Documents".

0040:err:shell:SHGetFolderPathAndSubDirW Failed to create directory L"C:\\users\\wineuser\\Desktop".

0040:err:ole:StdMarshalImpl_MarshalInterface Failed to create ifstub, hr 0x80004002

0040:err:ole:CoMarshalInterface Failed to marshal the interface {6d5140c1-7436-11ce-8034-00aa006009fa}, hr 0x80004002

0040:err:ole:apartment_get_local_server_stream Failed: 0x80004002

0024:err:shell:SHGetFolderPathAndSubDirW Failed to create directory L"C:\\users\\wineuser\\Local Settings\\Temporary Internet Files".

0024:err:wininet:cache_containers_init Couldn't get path for default container 0
0024:err:shell:SHGetFolderPathAndSubDirW Failed to create directory L"C:\\users\\wineuser\\Local Settings\\History".

0024:err:wininet:cache_containers_init Couldn't get path for default container 1
0024:err:shell:SHGetFolderPathAndSubDirW Failed to create directory L"C:\\users\\wineuser\\Cookies".

0024:err:wininet:cache_containers_init Couldn't get path for default container 2
Failed to read: session.screen0.titlebar.left

Setting default value
Failed to read: session.screen0.titlebar.right

Setting default value`

Viksie commented:
Having updated to version 5.3.3.4, the disappearing clips seems resolved automatically.

jshridha commented:
@Viksie Good to hear!

I'm merging this PR. Thanks to EVERYONE for all the hard work and wonderful contributions. Especially thanks to @leonowski for the PR and @TonyBrobston  for providing invaluable testing/debugging.

Builds should hit docker hub soon.

Viksie commented:
A year later and we're back on this topic.

With BlueIris now supporting Deepstack, you can easily send a AI alert image directly to Telegram with cURL. 

![image](https://user-images.githubusercontent.com/22875530/144422758-66903c34-1b98-4532-817d-bbeada4eb4c0.png)

The macros' values can be validated on a Windows instance with an Alert Action of "msg * /time:15 ALERT_PATH=&ALERT_PATH, ALERT_DB=&ALERT_DB, CAM=&CAM".

![image](https://user-images.githubusercontent.com/22875530/144420883-e53b380e-88ce-4728-8067-801f9f3b4098.png)

The resulting cURL parameter (on a Windows instance) to generate the above being: 

https://api.telegram.org/botXXXX/sendPhoto -F chat_id=XXXX -F caption="[*&CAM*]\(&WAN/alerts/&ALERT_DB?fulljpeg\) [UI3]\(&WAN/ui3.htm?rec=&ALERT_DB&maximize=1\) \- &MEMO" -F parse_mode=MarkdownV2 -F photo=@"C:\BlueIris\Alerts\&ALERT_PATH"

The first hyperlink showing the image and the second taking you directly to the action in a live playback via the web browser (UI3) interface.

With the Linux cURL the image seems to be delivered much faster than with Windows.

Yet, on the docker image, I cannot type the parameter (as amended) as the "(" and ")" characters does not generate. 

"https://api.telegram.org/botXXXX/sendPhoto -F chat_id=XXXX -F caption="[*&CAM*]\(&WAN/alerts/&ALERT_DB?fulljpeg\) [UI3]\(&WAN/ui3.htm?rec=&ALERT_DB&maximize=1\) \- &MEMO" -F parse_mode=MarkdownV2
-F photo=@"/home/wineuser/prefix/drive_c/Blue Iris/Alerts/&ALERT_PATH"" 

Neither can the text be copied over the browser. The " ( " and " ) " characters can be typed within a shell though. The only half baked "work-around" is to echo it into a file, but then it only creates more issues in trying to validate new issues....

So I hope @leonowski @jshridha @TonyBrobston @luck-y13 and @johnny2678 are all keen in resolving the "(" and ")" not generating.
